### PR TITLE
Improves proof coverage for aws_cryptosdk_*_header proofs

### PR DESCRIPTION
**Pull request by tegansb**
*Issue #, if available:* #543 

*Description of changes:* 
Addresses the low coverage of the aws_cryptosdk_sign_header and aws_cryptosdk_verify_header proofs. 
Includes a README for both the aws_cryptosdk_sign_header and aws_cryptosdk_verify_header proofs specifying expected coverage behavior. 
Replaces if (...) { abort() } with AWS_FATAL_POSTCONDITION(...), addressing Issue #543. 
Uses IMPLIES macro in the openssl stub for EVP_CIPHER_CTX_ctrl, which improves coverage. 

Coverage for aws_cryptosdk_sign_header improved from .8 to .92. 
Coverage for aws_cryptosdk_verify_header improved from .86 to .91. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c [hunk 1]
```diff
@@ -41,7 +41,8 @@ void aws_cryptosdk_decrypt_body_harness() {
 
     struct content_key *content_key;
 
-    uint8_t *tag = can_fail_malloc(props->tag_len);
+    uint8_t *tag = malloc(props->tag_len);
+    // possible that more memory could be allocated, but need to be able to write tag_len bytes to buffer tag.
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c:45]: Let's avoid C++ comments. Repeat this pattern in the other files as well.
```c
/*
 * Possible that more memory could be allocated, but need
 * to be able to write tag_len bytes to buffer tag.
 */
uint8_t *tag = malloc(props->tag_len);
```
Also, what do you mean by "but need to be able to write tag_len bytes to buffer tag"? Could you clarify?

File: .cbmc-batch/jobs/aws_cryptosdk_encrypt_body/aws_cryptosdk_encrypt_body_harness.c [hunk 2]
```diff
@@ -41,7 +41,8 @@ void aws_cryptosdk_encrypt_body_harness() {
 
     struct content_key *content_key;
 
-    uint8_t *tag = can_fail_malloc(props->tag_len);
+    uint8_t *tag = malloc(props->tag_len);
+    // possible that more memory could be allocated, but need to be able to write tag_len bytes to buffer tag.
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_encrypt_body/aws_cryptosdk_encrypt_body_harness.c:45]: See previous comment.

File: .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+0.92 (132 lines out of 144 statically-reachable lines in 23 functions reached)
+0.92 (132 lines out of 144 statically-reachable lines in 23 statically-reachable functions)
+
+## Expected Functions with Incomplete Coverage 
+
+(8/9) assert_byte_buf_equivalence: lhs==rhs never holds. 
+
+(8/9) aws_cryptosdk_alg_properties_is_valid: std_alg_props is never NULL
+
+(7/17) EVP_EncryptUpdate: Variable out is always NULL, so only that function behavior is explored.
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md:12]: Could you follow the same style we used for [MQTT proofs](https://github.com/aws/aws-iot-device-sdk-embedded-C/tree/v4_beta/cbmc/proofs/IotMqtt_PublishAsync)?

File: .cbmc-batch/jobs/aws_cryptosdk_verify_header/README.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+0.91 (136 lines out of 149 statically-reachable lines in 25 functions reached)
+0.91 (136 lines out of 149 statically-reachable lines in 25 statically-reachable functions)
+
+## Expected Functions with Incomplete Coverage 
+
+(8/9) assert_byte_buf_equivalence: lhs==rhs never holds. 
+
+(8/9) aws_cryptosdk_alg_properties_is_valid: std_alg_props is never NULL
+
+(7/17) EVP_DecryptUpdate: Variable out is always NULL, so only that function behavior is explored.
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_verify_header/README.md:12]: See previous comment about README style.

File: source/cipher.c [hunk 5]
```diff
@@ -234,7 +229,7 @@ int aws_cryptosdk_sign_header(
     if (!ctx) goto out;
 
     int outlen;
-    if (!EVP_CipherUpdate(ctx, NULL, &outlen, header->buffer, header->len)) goto out;
+    if (!EVP_EncryptUpdate(ctx, NULL, &outlen, header->buffer, header->len)) goto out;
```
Comment (feliperodri) [inline source/cipher.c:232]: @robin-aws @ttjsu-aws @danielsn is this valid?

File: source/cipher.c [hunk 6]
```diff
@@ -166,9 +166,7 @@ static int evp_gcm_encrypt_final(const struct aws_cryptosdk_alg_properties *prop
         return AWS_CRYPTOSDK_ERR_CRYPTO_UNKNOWN;
     }
 
-    if (outlen != 0) {
-        abort();  // wrong output size - potentially smashed stack
-    }
+    AWS_FATAL_POSTCONDITION(outlen == 0);
```
Comment (feliperodri) [inline source/cipher.c:169]: Nice!
```c
/* Wrong output size - potentially smashed stack. */
AWS_FATAL_POSTCONDITION(outlen == 0);
```

File: source/cipher.c [hunk 7]
```diff
@@ -203,10 +201,7 @@ static int evp_gcm_decrypt_final(
         }
         return AWS_CRYPTOSDK_ERR_CRYPTO_UNKNOWN;
     }
-
-    if (outlen != 0) {
-        abort();  // wrong output size - potentially smashed stack
-    }
+    AWS_FATAL_POSTCONDITION(outlen == 0);  // wrong output size - potentially smashed stack
```
Comment (feliperodri) [inline source/cipher.c:204]: Nice!
```c
/* Wrong output size - potentially smashed stack. */
AWS_FATAL_POSTCONDITION(outlen == 0);
```

File: .cbmc-batch/source/openssl/evp_override.c [hunk 8]
```diff
@@ -505,15 +505,15 @@ int EVP_CIPHER_CTX_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr) {
...
-        AWS_MEM_IS_WRITABLE(ptr, arg);  // need to be able to write taglen (arg) bytes to buffer ptr.
-    }
+    assert(IMPLIES(type == EVP_CTRL_GCM_GET_TAG, ctx->encrypt == 1));  // only legal when encrypting data
+    assert(IMPLIES(type == EVP_CTRL_GCM_GET_TAG, ctx->data_processed == true));
+    // need to be able to write taglen (arg) bytes to buffer ptr.
+    assert(IMPLIES(type == EVP_CTRL_GCM_GET_TAG, AWS_MEM_IS_WRITABLE(ptr, arg)));
+
+    assert(IMPLIES(type == EVP_CTRL_GCM_SET_TAG, ctx->encrypt == 0));  // only legal when decrypting data
+    // need to be able to write taglen (arg) bytes to buffer ptr.
+    assert(IMPLIES(type == EVP_CTRL_GCM_SET_TAG, AWS_MEM_IS_WRITABLE(ptr, arg)));
```
Comment (feliperodri) [inline .cbmc-batch/source/openssl/evp_override.c:515]: Let's avoid C++ comments.
```c
    /* Only legal when encrypting data. */
    assert(IMPLIES(type == EVP_CTRL_GCM_GET_TAG, ctx->encrypt == 1));
    assert(IMPLIES(type == EVP_CTRL_GCM_GET_TAG, ctx->data_processed == true));
    /* Need to be able to write taglen (arg) bytes to buffer ptr. */
    assert(IMPLIES(type == EVP_CTRL_GCM_GET_TAG, AWS_MEM_IS_WRITABLE(ptr, arg)));

    /* Only legal when decrypting data. */
    assert(IMPLIES(type == EVP_CTRL_GCM_SET_TAG, ctx->encrypt == 0));
    /* Need to be able to write taglen (arg) bytes to buffer ptr. */
    assert(IMPLIES(type == EVP_CTRL_GCM_SET_TAG, AWS_MEM_IS_WRITABLE(ptr, arg)));
```

File: .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c (hunk 1 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c]: Yes, I've made the comment clearer and changed the style as suggested.

File: .cbmc-batch/jobs/aws_cryptosdk_encrypt_body/aws_cryptosdk_encrypt_body_harness.c (hunk 2 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_encrypt_body/aws_cryptosdk_encrypt_body_harness.c]: Same as above

File: .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md (hunk 3 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md]: Done!

File: .cbmc-batch/jobs/aws_cryptosdk_verify_header/README.md (hunk 4 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_verify_header/README.md]: Done!

File: source/cipher.c (hunk 6 above)
Comment (tegansb) [reply source/cipher.c:169]: Thanks!

File: .cbmc-batch/source/openssl/evp_override.c (hunk 8 above)
Comment (tegansb) [reply .cbmc-batch/source/openssl/evp_override.c]: Yes, I've changed the comment style!

File: .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c [hunk 9]
```diff
@@ -41,7 +41,8 @@ void aws_cryptosdk_decrypt_body_harness() {
 
     struct content_key *content_key;
 
-    uint8_t *tag = can_fail_malloc(props->tag_len);
+    /* Need to allocate tag_len bytes for writing the tag */
+    uint8_t *tag = malloc(props->tag_len);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c:45]: `can_fail_malloc()` followed by `assume != null` is better documentation.
also, how do we know its safe to assume nonnull?

File: .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md [hunk 10]
```diff
@@ -0,0 +1,18 @@
+# Memory safety proof for aws_cryptosdk_sign_header
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md:1]: Standard question: is this intrinsic to the harness, or the function?

File: .cbmc-batch/jobs/aws_cryptosdk_sign_header/aws_cryptosdk_sign_header_harness.c [hunk 11]
```diff
@@ -28,10 +28,7 @@ void aws_cryptosdk_sign_header_harness() {
     struct aws_byte_buf header;
 
     /* assumptions*/
-    __CPROVER_assume(props);
-    __CPROVER_assume(
-        props->impl->cipher_ctor == EVP_aes_128_gcm || props->impl->cipher_ctor == EVP_aes_192_gcm ||
-        props->impl->cipher_ctor == EVP_aes_256_gcm);
+    __CPROVER_assume(aws_cryptosdk_alg_properties_is_valid(props));
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_sign_header/aws_cryptosdk_sign_header_harness.c:31]: much nicer

Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_sign_header/aws_cryptosdk_sign_header_harness.c:31]: has this been added as a precondition on the function?

Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_sign_header/aws_cryptosdk_sign_header_harness.c:31]: It hadn't been but I've gone ahead and added it as a precondition for both aws_cryptosdk_sign_header and aws_cryptosdk_verify_header.

File: .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md (hunk 10 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md:1]: The function -- I've added additional documentation to both READMEs.

File: .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c (hunk 9 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c:45]: If the tag is NULL, then the assert in EVP_CIPHER_CTX_ctrl that taglen bytes can be written to buffer tag fails. I realize that there is no corresponding precondition on decrypt_body to ensure this though so I have gone ahead and added one. The analogous applied for encrypt_body.

File: .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c [hunk 12]
```diff
@@ -41,7 +41,9 @@ void aws_cryptosdk_decrypt_body_harness() {
 
     struct content_key *content_key;
 
+    /* Need to allocate tag_len bytes for writing the tag */
     uint8_t *tag = can_fail_malloc(props->tag_len);
+    __CPROVER_assume(tag != NULL);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c:46]: is this a precondition on the function

File: .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+This proof harness attains 92% code coverage.  The following comments explain
+why the uncovered lines of code are unreachable code.
+
+Some functions contain unreachable blocks of code:
+
+* `assert_byte_buf_equivalence`
+
+    * lhs==rhs never holds.
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md:10]: why?

File: source/cipher.c (hunk 6 above)
Comment (danielsn) [inline source/cipher.c:169]: add the comment here explaining why

File: .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c (hunk 12 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c:46]: It wasn't but I added the more refined and required AWS_MEM_IS_WRITABLE as a check in a previous commit.

File: source/cipher.c (hunk 6 above)
Comment (tegansb) [reply source/cipher.c:169]: Done

File: .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md (hunk 13 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_sign_header/README.md]: Added.

File: .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c (hunk 12 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_decrypt_body/aws_cryptosdk_decrypt_body_harness.c:46]: Good

File: .cbmc-batch/jobs/aws_cryptosdk_encrypt_body/aws_cryptosdk_encrypt_body_harness.c [hunk 14]
```diff
@@ -41,7 +41,9 @@ void aws_cryptosdk_encrypt_body_harness() {
 
     struct content_key *content_key;
 
+    /* Need to allocate tag_len bytes for writing the tag */
     uint8_t *tag = can_fail_malloc(props->tag_len);
+    __CPROVER_assume(tag != NULL);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_encrypt_body/aws_cryptosdk_encrypt_body_harness.c:46]: Is this a precondition on the function?

**Review by danielsn — APPROVED** [review]

**Pull request merged**